keps/369-job-interface/README.md

```diff
@@ -0,0 +1,329 @@
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+We collected feedbacks from the community about how to fully integrate Kueue with MPIJob,
+see [#499](https://github.com/kubernetes-sigs/kueue/issues/499).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- Job interface here is more like a template and tell developers what to do, but even without
+implementing the interface people can also build their controllers.
+- This will increase the code complexity by wrapping the original Jobs.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Part I: Job Interface
+
+We will define a new interface named Job, this should be implemented by custom job-like applications:
+
+```golang
+type Job interface {
+  // Suspend instructs whether the job is suspended or not.
+  Suspend() bool
+  // Start means starting the job, like unsuspending the job.
+  Start(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // Stop means stopping the job, like suspending the job.
+  Stop(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // InjectNodeAffinity will inject the node affinity extracting from workload to job.
+  InjectNodeAffinity(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // RestoreNodeAffinity will restore the original node affinity of job.
+  RestoreNodeAffinity(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // Finished instructs whether the job is completed/failed or not.
+  Finished() bool
+  // Finish will update the workload to finished.
+  Finish(ctx context.Context, client client.Client, wl *kueue.Workload) error
+  // EquivalentToWorkload validates whether the workload is semantically equal to the job.
+  EquivalentToWorkload(wl kueue.Workload) bool
+  // CreateWorkload will create a workload constructing by the job.
+  CreateWorkload(ctx context.Context, client client.Client, scheme *runtime.Scheme) (*kueue.Workload, error)
```

Comment:
Maybe this should be something like `PodSets` that will return `[]kueue.PodSet`. Then from there we can provide code to create the Workload object.

Reply:
We can for sure obtain the namespace from the ObjectMeta. I wonder if there are any APIs that already include a queue name in a field. Otherwise we can also obtain the queue name from the (standard) annotation in the ObjectMeta.

Reply:
Fine, then we'll have a help function named like `createWorkload()` with client passed-in.